scripts/ai_determine_version_bump.py

```diff
@@ -0,0 +1,564 @@
+- CLAUDE_MODEL (default: claude-3-5-sonnet-20241022) - specify which Claude model to use
+- GEMINI_MODEL (default: gemini-2.0-flash-exp) - specify which Gemini model to use
+
+The script will try providers in order: OpenAI -> Claude -> Gemini -> Fallback
+
+Install: pip install openai anthropic google-generativeai
+"""
+
+import argparse
+import json
+import os
+import re
+import subprocess
+import sys
+from typing import Tuple
+
+
+def get_latest_tag() -> str:
+    """Get the latest git tag."""
+    try:
+        result = subprocess.run(
+            ["git", "tag", "--sort=-v:refname"],
+            capture_output=True,
+            text=True,
+            check=True,
+        )
+        tags = [
+            line.strip() for line in result.stdout.strip().split("\n") if line.strip()
+        ]
+        if tags:
+            return tags[0]
+        return ""
+    except subprocess.CalledProcessError:
+        return ""
+
+
+def get_commits_since_tag(tag: str) -> list[dict]:
+    """Get commit messages and diffs since the given tag."""
+    try:
+        if tag:
+            cmd = [
+                "git",
+                "log",
+                f"{tag}..HEAD",
+                "--pretty=format:%H|||%s|||%b",
+            ]
+        else:
+            cmd = ["git", "log", "--pretty=format:%H|||%s|||%b"]
+
+        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
+
+        commits = []
+        for line in result.stdout.strip().split("\n"):
+            if not line.strip():
+                continue
+            parts = line.split("|||")
+            if len(parts) >= 2:
+                commit_hash = parts[0]
+                subject = parts[1]
+                body = parts[2] if len(parts) > 2 else ""
+
+                # Get file changes for this commit
+                diff_cmd = ["git", "diff", f"{commit_hash}^..{commit_hash}", "--stat"]
+                diff_result = subprocess.run(diff_cmd, capture_output=True, text=True)
+                files_changed = diff_result.stdout.strip()
+
+                commits.append(
+                    {
+                        "hash": commit_hash[:7],
+                        "subject": subject,
+                        "body": body,
+                        "files_changed": files_changed,
+                    }
+                )
+
+        return commits
+    except subprocess.CalledProcessError:
+        return []
+
+
+def analyze_with_openai(commits_summary: str, current_version: str) -> dict:
+    """Use OpenAI to analyze commits."""
+    api_key = os.getenv("OPENAI_API_KEY")
+    if not api_key:
+        raise ValueError("OPENAI_API_KEY not set")
+
+    model = os.getenv("OPENAI_MODEL", "gpt-4o")
+
+    from openai import OpenAI
+
+    client = OpenAI(api_key=api_key)
+
+    prompt = f"""You are analyzing git commits for a CUDA kernel library called FlashInfer to determine the appropriate semantic version bump.
```

Comment:
This prompt is duplicated across the different models. I would suggest to assign it to a variable, or put it into a separate file.